packages/model/addon/-private/system/many-array.js

```diff
@@ -140,39 +136,33 @@ export default EmberObject.extend(MutableArray, DeprecatedEvented, {
     @private
     */
     this.currentState = [];
-    this.flushCanonical(this.initialState, false);
-    // we don't need this anymore, it just prevents garbage collection the records in the initialState
-    this.initialState = undefined;
+    this._isUpdating = false;
+    this._isDirty = false;
+    // make sure we initialize to the correct state
+    // since the user has already accessed
+    this.retrieveLatest();
   },
 
-  // TODO: if(DEBUG)
-  anyUnloaded() {
-    // Use `filter[0]` as opposed to `find` because of IE11
-    let unloaded = this.currentState.filter((im) => im._isDematerializing || !im.currentState.isLoaded)[0];
-    return !!unloaded;
+  // TODO refactor away _hasArrayObservers for tests
+  get _hasArrayObservers() {
+    return this.hasArrayObservers || this.__hasArrayObservers;
   },
 
-  removeUnloadedInternalModel() {
-    for (let i = 0; i < this.currentState.length; ++i) {
-      let internalModel = this.currentState[i];
-      let shouldRemove;
-      if (CUSTOM_MODEL_CLASS) {
-        shouldRemove = internalModel._isDematerializing;
-      } else {
-        shouldRemove = internalModel._isDematerializing || !internalModel.currentState.isLoaded;
-      }
-      if (shouldRemove) {
-        this.arrayContentWillChange(i, 1, 0);
-        this.currentState.splice(i, 1);
-        this.set('length', this.currentState.length);
-        this.arrayContentDidChange(i, 1, 0);
-        return true;
-      }
+  notify() {
+    this._isDirty = true;
+    if (this._hasArrayObservers) {
+      this.retrieveLatest();
+    } else {
+      this.notifyPropertyChange('[]');
+      this.notifyPropertyChange('firstObject');
+      this.notifyPropertyChange('lastObject');
```

Comment:
Would you mind explaining these two blocks? What are we trying to accomplish with the non array observers block?  What will follow in this case due to the `notifyPropertyChange` on `[]`, `firstObject` and `lastObject`?

Reply:
retrieveLatest is the syncFlush. If we have observers we must flush immediately. If we don't have any we can notify and wait for any subscribers to pull on the ManyArray content to flush.

In the observers case observers are notified by arrayContentWillChange and arrayContentDidChange